pkg/chains/signing.go

```diff
@@ -58,21 +62,47 @@ type TaskRunSigner struct {
 }
 
 // IsSigned determines whether a TaskRun has already been signed.
-func IsSigned(tr *v1beta1.TaskRun) bool {
+func IsSigned(tr *v1beta1.TaskRun) (bool, string) {
+	const signed = "true"
 	signature, ok := tr.ObjectMeta.Annotations[ChainsAnnotation]
 	if !ok {
-		return false
+		return false, ""
 	}
-	return signature == "true"
+	return signature == signed, signature
+}
+
+// returns true if:
+// the signed annotation isn't there, the failed annotation isn't there, and retries<3
+func RetryAgain(tr *v1beta1.TaskRun) (bool, string) {
+	retries, ok := tr.ObjectMeta.Annotations[ChainsRetriesAnnotation]
+	if !ok {
+		return false, ""
+	}
+	return true, retries
 }
 
 // MarkSigned marks a TaskRun as signed.
-func MarkSigned(tr *v1beta1.TaskRun, ps versioned.Interface, annotations map[string]string) error {
+func MarkSigned(tr *v1beta1.TaskRun, ps versioned.Interface, annotations map[string]string,
+	state string) error {
+	return markannotation(tr, ps, annotations, ChainsAnnotation, state)
+}
+
+// AddRetries marks a TaskRun as retrying.
+func AddRetries(tr *v1beta1.TaskRun, ps versioned.Interface, annotations map[string]string,
+	value int) error {
```

Comment:
As i mentioned in the above comment, we can remove `value int` from the header of this function, and read it from the annotations within the function

Reply:
Took care of it. Thanks